Assignment-Week2-ibrahim

ImAltay opened the pull request:
https://666c0a579758259b92d18812--playful-banoffee-1d6760.netlify.app/


14/06/2024 - 11:17 CEST
### Missing requirements:
-   App needs to show an error message if the request failed

tlorent commented on week2/project/ecommerce/fake-data/all-categories.js:
The fake data can be removed now that you are working with an API 😄

tlorent commented on week2/project/ecommerce/fake-data/all-products.js:
The fake data can be removed now that you are working with an API 😄

tlorent commented on week2/project/ecommerce/src/App.jsx:
```diff
@@ -0,0 +1,87 @@
+  const [allCategories, setAllCategories] = useState([]);
+  const [productsList, setProductsList] = useState([]);
+  const [loading, setLoading] = useState(true);
+  // const [productId, setProductId] = useState(-1);
```
```suggestion
```

tlorent commented on week2/project/ecommerce/src/App.jsx:
```diff
@@ -0,0 +1,87 @@
+    }
+  };
+
+  // const handleProductClick = (id) => {
+  //   setProductId(id);
+   
+  // };
```
```suggestion
```

tlorent commented on week2/project/ecommerce/src/App.jsx:
```diff
@@ -0,0 +1,87 @@
+                productsList={productsList}
+                loading={loading}
+                activeCategory={activeCategory}
+                // handleClick = {handleProductClick}
```
```suggestion
```

tlorent commented on week2/project/ecommerce/src/App.jsx:
```diff
@@ -0,0 +1,87 @@
+
+export default App;
+
+// <div>
+// <Categories categoriesList={allCategories} handleClick={handleClick} />
+// <Products productsList={productsList} loading={loading} activeCategory={activeCategory} />
+// </div>
```
```suggestion
```

tlorent commented on week2/project/ecommerce/src/components/ProductInfo.jsx:
```diff
@@ -0,0 +1,31 @@
+
+const Product = () => {
+    const { id } = useParams();
+    // 
```
```suggestion
```

tlorent commented on week2/project/ecommerce/src/components/ProductInfo.jsx:
```diff
@@ -0,0 +1,31 @@
+const Product = () => {
+    const { id } = useParams();
+    // 
+    const [productTitle, setProductTitle ] = useState('');
+    const [productDescription, setProductDescription ] = useState('');
+    const [productImgUrl, setProductImgUrl ] = useState('');
+
+
+
+    useEffect(() => {
+        fetch(`https://fakestoreapi.com/products/${id}`)
+            .then((res) => res.json())
+            .then((data) => {
+                setProductTitle(data.title);
+                setProductDescription(data.description);
+                setProductImgUrl(data.image);
+            });
+    });
```
Maybe it's useful to put all of this state into one object? So you don't need multiple state updates, multiple variables to keep track of and then you can just store it all in one object because they are all related to a product? 🙂

tlorent commented on week2/project/ecommerce/src/components/ProductInfo.jsx:
```diff
@@ -0,0 +1,31 @@
+    return (
+        <div className='product-info'>
+            <h2>{productTitle}</h2>
+            <img src={productImgUrl} alt="" />
```
Great that you added an `alt` tag! However, it is empty now and in the case of an image like this it would actually be beneficial to add a description. Can you please add that? 🙂 Maybe the product's description?

tlorent commented on week2/project/ecommerce/src/App.jsx:
```diff
@@ -0,0 +1,87 @@
+  // };
+
+  // fetch data on mount for all categories
+  useEffect(() => {
+    fetch('https://fakestoreapi.com/products/categories')
+      .then((res) => res.json())
+      .then((data) => {
+        setAllCategories(data);
+      });
+  }, []);
+  // fetch data on mount for all or selected products
+  useEffect(() => {
+    setLoading(true);
+    if (activeCategory === '') {
+      fetch('https://fakestoreapi.com/products')
+        .then((res) => res.json())
+        .then((data) => {
+          setProductsList(data);
+          setLoading(false);
+        });
+    } else {
+      fetch(`https://fakestoreapi.com/products/category/${activeCategory}`)
+        .then((res) => res.json())
+        .then((data) => {
+          setProductsList(data);
+          setLoading(false);
+        });
+    }
+  }, [activeCategory]);
```
Great to see you're working with hooks! 🙌 Maybe you can try using `async/await` here because it's a more modern approach and easier to read + maintain?

ImAltay replied:
I converted to this method. but now I get eslint warning:

` React Hook useEffect has a missing dependency: 'fetchProducts'. Either include it or remove the dependency array. `

So I did what it want me to do then it also want me to put it in useCallback. which I did too and app works fine(and was working fine with eslint warning too) . but I do not understand why we did this. I also dont understand why it wanted me to utilize useCallback with fetchProducts() but not also wth fetchCategories();

tlorent replied:
@ImAltay Thank you for your patience, I was on holiday for two weeks so I could not reply. The dependency array of `useEffect` is needed because this indicates when the effect should run again, when the value(s) inside the array changes. In this case we are using `fetchProducts` inside the effect so when that function changes the effect should run again. However, because React creates your component again including the functions inside it (it will be a new reference every time the component renders), this means `useEffect` will run **every render** 😲 In this case this is not something we want, so that is why we wrap it in `useCallback` to _"memoize"_ it 🙂 This basically means that React will create a memorized version of the function so that it knows it does/will not change and the effect will now not re-run based on that value in the dependency array.

Another thing you can potentially do is not include it in the dependency array and add a comment why, but in this case I would go for the lint recommendation 👍

tlorent reviewed (changes requested):
Nice work @ImAltay! 🙌 Your deployed code looks good, site works really well. Some comments:

1. As a general rule, it is good to delete commented out code 🙂 If it's not relevant anymore then it is better to get rid of it to not confuse other developers and to not ship unnecessary code.
2. I think the data fetching can be improved a bit by using `async/await`

ImAltay commented:
Thank you for the feedback.

I was quite behind of the deadline. I totally didnt even realize I have commented out codes 🙈 
I will implement the changes ASAP

ImAltay commented:
I implemented the changes. I also added errors. but I kinda went with my own flow with those.  and I am not quite sure if showing errors this way is the best

tlorent reviewed (approved):
🎉
